### ISSUE TITLE
Loss not decreasing when training NER to recognize new entity

### ISSUE THREAD

**Issue by MkdmOmar** [explosion/spaCy#1140]
I went through the tutorial on adding an 'ANIMAL' entity to spaCy NER [here](https://github.com/explosion/spacy/blob/master/examples/training/train_new_entity_type.py).

The tutorial only includes 5 sentences, which is obviously nowhere near enough to rigorously train the NER. So I manually retrieved text from the internet on 25 different animals. I went through each document and annotated the occurrences of every animal. Then I wrote a Python script that goes through all 25 files and build a large trainingData array of (doc, (start_char, end_char, label)) tuples. The script adds 'ANIMAL' tuples for every annotation I marked. It also retrieves all the entities that spaCy NER recognizes and adds those entities whose [start, end] range does not intersect with my 'ANIMAL' entities. I add these entities to avoid the 'forgetting' problem described in the tutorial. The final trainingData array had 573 unique sentences (docs) and 1279 unique entity labels. 

I then execute`nlp.entity.add_label('ANIMAL')` and run the trainer. However, the loss simply fluctuates between ~850 and ~1200 and does not seem to decrease beyond that range, even after having run the trainer for more than 30,000 iterations.

The only difference between my `train_ner()` and the tutorial version is that I commented out the lines:
`random.seed(0)`
`nlp.entity.model.learn_rate = 0.001`

I also iterate my training for a million iterations instead of a thousand since I have more data and the loss does not seem to be decreasing (even after 30k iterations).

Am I doing something wrong here? The tutorial example (with only 5 sentences) converges very quickly in around ~150 iterations. Will adding a lot of examples (573 sentences in my case) increase the training time by a considerable amount (training for 5 hours and counting, yet no apparent decrease in loss in my case)?

## My Environment

**Info about spaCy**

* **Python version:** 2.7.12
* **Platform:** Linux-4.4.0-57-generic-x86_64-with-Ubuntu-16.04-xenial
* **spaCy version:** 1.8.2
* **Installed models:**

**Comment by honnibal** [comment]
I think the learning rate may be your problem here. If the learning rate is too high, SGD may not converge. Did the script fail somehow when you left it at `learn_rate = 0.001`?

**Comment by MkdmOmar** [comment]
The script does not fail if I leave in the learning rate - it just gives the same behavior I described above. I also tried larger learning rates (0.005, 0.01, and 0.05), but that simply leads to larger losses. I also tried smaller learning rates (0.0005 and 0.0001), but I still get the same large losses (in the 850 - 1150 range) and they don't seem to be decreasing beyond that range. I have not run these trials for more than 1000 iterations, however.

**Comment by honnibal** [comment]
If the loss isn't decreasing, doing more of the same is unlikely to help.

Have you verified that your data is correct? It's easy to make an off-by-one error in the start/end indices.

**Comment by MkdmOmar** [comment]
I have examined my data and it appears to be correct. Here are two randomly chosen elements in my trainingData list:

`(u"The largest species of gecko is the Delcourt's gecko (which is now believed to be extinct), native to New Zealand and it grows to nearly 60 cm in length.",
  [(23, 28, u'ANIMAL'),
   (47, 52, u'ANIMAL'),
   (36, 44, u'PERSON'),
   (102, 113, u'GPE'),
   (130, 139, u'CARDINAL')]),`

` (u'The female starfish is capable is releasing over 2 million eggs at any one time, although the average amount of eggs that the female starfish releases is closer to 1 million.',
  [(11, 19, u'ANIMAL'),
   (133, 141, u'ANIMAL'),
   (44, 63, u'MONEY'),
   (71, 74, u'CARDINAL'),
   (164, 173, u'CARDINAL')]),`

I have 573 of these.

**Comment by MkdmOmar** [comment]
I also noticed something in the [tutorial code](https://github.com/explosion/spaCy/blob/master/examples/training/train_new_entity_type.py) that puzzled me. Specifically, this block of code:

```      
        for raw_text, entity_offsets in train_data:
            gold = GoldParse(doc, entities=entity_offsets)
            # By default, the GoldParse class assumes that the entities
            # described by offset are complete, and all other words should
            # have the tag 'O'. You can tell it to make no assumptions
            # about the tag of a word by giving it the tag '-'.
            # However, this allows a trivial solution to the current
            # learning problem: if words are either 'any tag' or 'ANIMAL',
            # the model can learn that all words can be tagged 'ANIMAL'.
            #for i in range(len(gold.ner)):
                #if not gold.ner[i].endswith('ANIMAL'):
                #    gold.ner[i] = '-'
            doc = nlp.make_doc(raw_text)  
```

In the code block above, GoldParse() takes in the doc that was initially defined in this previous loop on line 44 (shown below):


```
    # Add new words to vocab
    for raw_text, _ in train_data:
        doc = nlp.make_doc(raw_text)
        for word in doc:
            _ = nlp.vocab[word.orth]
```
That doc will not have the text that corresponds to the entity_offsets in the iteration of the current loop (`for raw_text, entity_offsets in train_data`). So I was wondering why doc is redefined with the relevant raw_text only after GoldParse is called, since GoldParse will take in a doc that does not contain the current entities.

So I moved the line `doc = nlp.make_doc(raw_text)` directly above the line `gold = GoldParse(doc, entities=entity_offsets)` to address this observation and now the loss decreases significantly well. At iteration 0 the loss is 1084. By iteration 7, the loss is 4. It converged to a loss of 1-2 by the 10th iteration and has remained that way (now at 2k iterations).

Does this fix sound reasonable and is there a reason why `doc = nlp.make_doc(raw_text)` was placed after `gold = GoldParse(doc, entities=entity_offsets)` ?

**Comment by honnibal** [comment]
I think that's definitely just a bug in the tutorial --- thanks for getting to the bottom of this!

**Comment by ankur7721** [comment]
@MkdmOmar Can you please tell me how you are observing Training loss, what are the code changes you did [here](https://github.com/explosion/spacy/blob/master/examples/training/train_new_entity_type.py)
and did you made any graph for training loss?

**Comment by MkdmOmar** [comment]
@ankur7721 the loss is defined on line 53 `loss = 0.` and is updated across all training examples on line 71 `loss += nlp.entity.update(doc, gold, drop=0.9)`, for every iteration. I removed the drop option since it's not supported in my version of spaCy (1.8.2).

I used matplotlib.pyplot to plot the training loss over 50 iterations. To do this I just made a list called losses and appended the updated loss to losses just before line 72. Then I plot the losses using matplotlib at the end of `train_ner()`. Below is my loss plot:
![loss](https://user-images.githubusercontent.com/13189476/27391798-e3ea6006-5672-11e7-84a4-13516609a1c5.PNG)

**Comment by benolayinka** [comment]
The [train_new_entity_type.py](https://github.com/explosion/spacy/blob/master/examples/training/train_new_entity_type.py) is slightly newer and includes some features that only work in spaCy 1.9, like using the model learn rate and dropout rate. Is it reasonable to use this training process to update existing entity recognition as well?

The difference between train_new_entity_type and train_ner appears to be that in train_ner, an EntityRecognizer is instantiated manually for the nlp vocab and then trained. What is the difference between doing this, and simply calling nlp.entity.update on the existing model?
